### PR TITLE
perf(server): reuse local submodule objects when creating worktrees

### LINKED ISSUE #9853: [Bug]: New worktree creation reclones already-initialized submodule objects
### Before submitting

- [x] I searched existing issues and did not find a duplicate.
- [x] I included enough detail to reproduce or investigate the problem.

### Area

apps/server

### Steps to reproduce

1. Open a repository with an initialized submodule containing substantial content/history. The measured example has roughly 1 GiB of packed submodule objects already available in the source checkout.
2. Set new threads to **New worktree**, with **Start from origin** enabled.
3. Start a new thread and send its first message.
4. Inspect the server trace for `fetchRemote`, `GitVcsDriver.createWorktree`, and `GitVcsDriver.createWorktree.updateSubmodules`.

### Expected behavior

Preparing an independent worktree should reuse locally available submodule objects where possible, preserve the selected gitlink commit, and fetch objects that are missing. The new worktree should remain usable if the source submodule is subsequently removed or pruned.

### Actual behavior

The first message waits roughly 18 seconds for worktree preparation in the observed case. Most of that time is submodule initialization, not fetching origin or checking out the superproject.

| Existing server trace step | Duration |
|---|---:|
| Fetch origin | 0.522 s |
| `git worktree add` | 0.296 s |
| Recursive submodule initialization | 17.417 s |

Each linked worktree gets a separate submodule Git directory. The current `git submodule update --init --recursive` invocation clones objects again rather than explicitly using the populated source checkout as a reference. This happens before the first turn is dispatched. Turning off Start from origin would remove only about half a second in this example.

### Impact

Major degradation or frequent failure — a repeated startup delay when creating fresh worktree threads; the operation eventually succeeds.

### Version or commit

Source inspected and benchmarked at `8faf031c2dac4f147c43fe32e326b76678c20c94`.

### Environment

Linux, Git 2.55.0. Benchmarks used Node 26.8.1; focused verification of the proposed fix also passed on Node 24.20.0. The affected operation runs in the server's Git driver.

### Logs or stack traces

No error is required to reproduce this: the slow operations complete successfully. The relevant redacted timing breakdown is above.

### Supporting measurements

A locally tested candidate supplies initialized source submodules as temporary `--reference` repositories and uses `--dissociate` to avoid a persistent dependency on their object stores.

| Measurement | Original run 1 | Original run 2 | Candidate run 1 | Candidate run 2 |
|---|---:|---:|---:|---:|
| Recursive submodule command | 17.288 s | 17.440 s | 4.721 s | 6.574 s |
| Actual `createWorktree()` driver call | 26.089 s | 23.400 s | 5.721 s | 5.501 s |

The driver-call mean fell from 24.744 s to 5.611 s, approximately 77%. Driver measurements include worktree creation, reference discovery, submodule initialization, base-ref configuration and cache invalidation; they do not include the earlier origin fetch or provider startup. These are separate runs from the original server trace, so their absolute durations differ.

Each comparison used fresh, disk-backed disposable worktrees at the same submodule commit, with interleaved baseline/candidate ordering, the same remote, `pack.threads=2`, and process priority `nice 10`. There were two observations per variant per measurement on a shared machine, without flushing system caches. This is evidence for this workload, not a universal speedup claim.

Git Trace2 recorded about 14.84–14.94 s in `index-pack` for the original submodule commands, versus about 2.12–2.33 s in the candidate's local `repack`. Network byte savings were not measured directly. Exact submodule HEAD and the absence of persistent alternates were checked.

Focused tests found no correctness regression, including nested/multiple submodules, shallow or uninitialized sources, missing cached commits, and source object-store removal. Windows/macOS, large submodule counts, and performance with local-path remotes remain unmeasured. References still require remote contact; fallback after a failed reference attempt can add work.

### Workaround

Reuse an existing worktree when sharing its working directory is appropriate. Disabling Start from origin does not address the dominant submodule cost in this case.

### Related reports

#9340 covers a related blocking startup path and progress reporting, particularly with an upstream base. This report isolates repeated submodule object cloning. #7672 explains why automatic submodule initialization is needed; that behavior should be preserved.

### PR DESCRIPTION
**Pull request by Clowdyffs**
fixing since i do a lot of work on a repo with a decent amount of submodule content, so everytime i start a new thread (I have it start a new worktree based on origin for new threads), and send my first message, it would take ~18 seconds to setup the worktree, this does some work to cut that down. 

======= below this line is ai slop =========

## What Changed

Fresh worktrees clone submodule history again even when the source checkout already has the objects. Initialize each destination submodule using its matching source checkout as a temporary Git `--reference`, then `--dissociate` so the destination owns its objects. Repeat for nested submodules. Missing objects still come from the configured remote, and unsuitable references fall back to an ordinary clone.

Git's submodule wrapper retains only the last `--reference`, so each clone now receives exactly one matching reference. Custom submodule active/update policies and clone concurrency settings retain the ordinary recursive command. If preparation still fails, retry the ordinary recursive update so Git can process remaining siblings before the existing best-effort warning.

The production diff is 98 changed lines in the shared server Git driver; the only other changed file is its focused test suite. No new settings, persistent cache, dependencies, or client/protocol changes.

Closes #9853.

## Why

An observed first-message startup spent 0.522 s fetching origin, 0.296 s adding the worktree, and 17.417 s initializing submodules. This change targets the dominant submodule step while preserving automatic population and independent worktrees.

### Worktree setup measurements

**72.1% less time · 15.82 seconds saved on average**

| Implementation | Run 1 | Run 2 | Average | Time reduction |
|---|---:|---:|---:|---:|
| Before | 21.19 s | 22.71 s | **21.95 s** | **0%** |
| After | 6.20 s | 6.05 s | **6.13 s** | **72.1%** |

*Actual `createWorktree()` calls, including submodule initialization. Linux/btrfs · Git 2.55 · approximately 1 GiB of packed submodule history. Two runs per version, interleaved, using fresh destinations and the same pinned commit. Excludes origin fetch and provider startup; caches were not flushed.*

<details>
<summary>Measurement method and limits</summary>

These measurements were taken on `57a6e533d`, replacing the earlier implementation's measurements. The subsequent correction handles global `submodule.active=true` selection and failure fallback; it does not change the successful path exercised by this benchmark. Baseline: `8faf031c2dac4f147c43fe32e326b76678c20c94`. Node 26.8.1, disk-backed destinations, the same remote, `pack.threads=2`, and `nice 10`. Execution order was before, after, after, before. The baseline driver was loaded from the original source with only benchmark import paths adjusted.

The measured driver call includes worktree creation, submodule preparation, base-ref configuration and cache invalidation. Each run checked the exact submodule HEAD and absence of persistent alternates. The original ~18-second trace and these benchmarks are separate samples on a shared machine with network variability. Two observations per version establish a local improvement, not a universal speedup; network byte savings were not measured directly.

The large-repository timing above measures one large submodule. The separate edge-case matrix below measures multiple/nested submodules, custom policies, and fallback overhead on Linux/btrfs. Local-path remotes may already benefit from Git's local clone optimizations; the new fixtures use file:// transport to avoid hardlink cloning.

</details>

### Edge-case measurements on the final commit

The matrix below compares the original driver (`8faf031c`) with the unchanged PR head (`1001b4cc0`). **120 measured calls: 12 scenarios, five calls per implementation per scenario.** Destination state matched the original driver in every measured call.

| Scenario | Before, mean | After, mean | Time reduction |
|---|---:|---:|---:|
| 1 cached submodule | 122.5 ms | 100.5 ms | +17.9% |
| 4 cached sibling submodules | 410.4 ms | 325.5 ms | +20.7% |
| 2 parents + 2 nested submodules | 347.3 ms | 306.4 ms | +11.8% |
| 12 tiny cached submodules | 240.4 ms | 404.5 ms | -68.2% |
| Active pathspec: 2 of 4 selected | 205.3 ms | 193.1 ms | +5.9% |
| Literal active=true pathspec | 109.2 ms | 116.8 ms | -6.9% |
| update=none: 1 of 4 skipped | 267.5 ms | 284.9 ms | -6.5% |
| Custom fetchJobs=4 | 227.2 ms | 227.8 ms | -0.3% |
| Stale + dirty source, 4 modules | 319.1 ms | 268.0 ms | +16.0% |
| Shallow sources, 4 modules | 308.3 ms | 433.2 ms | -40.5% |
| Uninitialized sources, 4 modules | 327.1 ms | 371.5 ms | -13.6% |
| Unreachable first module + 4 healthy siblings | 326.5 ms | 347.7 ms | -6.5% |

*Positive percentages mean less elapsed time; negative percentages mean more. These are synthetic local `file://` transport fixtures on Linux/btrfs, separate from the approximately 1 GiB repository measurement above. Ordinary modules contain approximately 16 MiB of history each (four revisions of a 4 MiB random payload); the tiny modules contain a README only. Nested modules use spaces in their paths, and submodule names differ from their paths.*

The slower cases have small absolute costs in these fixtures, even where the percentage increase is large. The command paths and packet traces explain the likely sources of overhead; the timings measure the whole driver call, not the milliseconds attributable to each subprocess:

- **12 tiny cached modules: +164.1 ms (240.4 → 404.5 ms).** The optimized runs made zero object `want` requests, so reuse worked. Each module still requires a separate update invocation, reference setup, and dissociation to preserve independence. With only README-sized histories to transfer, this setup costs more than the transfer work it saves.
- **Four shallow sources: +124.9 ms (308.3 → 433.2 ms).** Git rejects shallow repositories as clone references, so the attempted reference clones retry as ordinary clones. The fixture shows no reduction in object requests; reference attempts and per-module setup add work. See [Git's reference validation](https://github.com/git/git/blob/v2.55.0/odb.c#L296-L300).
- **Four uninitialized sources: +44.4 ms (327.1 → 371.5 ms).** There are no initialized source checkouts to borrow from. The driver checks for references and invokes updates per module, but each module still needs ordinary cloning.
- **An unreachable first module: +21.1 ms (326.5 → 347.7 ms).** The per-module attempt fails before the driver retries the ordinary recursive command, which processes the healthy siblings. This fixture fails immediately at a local missing URL; a slow remote failure can make retries more costly.

Custom active/update/concurrency settings use Git's ordinary recursive path after the configuration checks. Their measured mean differences range from **12.2 ms less to 17.4 ms more**; the fetchJobs=4 difference is **0.6 ms more**, with substantial sample variation. Those small differences do not establish either an optimization or a precise overhead cost for custom policies.

For interactive first-message setup, the **44–164 ms** penalties in the tiny/shallow/uninitialized fixtures are likely minor: they are paid once when creating a worktree, not on every subsequent message in that worktree. This is a practical assessment, not a measured perceptibility result. The separate large-repository sample saved **15.82 seconds**; that is context for the tradeoff, not a promise that the same user receives both results. More modules or slower failures can increase the penalty, so the negative rows remain visible rather than being treated as universally negligible.

<details>
<summary>Method, behavioral checks, and every timing sample</summary>

Every sample runs the actual `createWorktree()` driver call, using a fresh disk-backed destination. Source fixtures are initialized before timing, and each scenario uses the same source and selected commit for both implementations. Execution order is before, after, after, before, before, after, after, before, before, after. Caches were not flushed; runs were serial on a shared machine, at `nice 10`, with `pack.threads=2`. Fixture setup, validation, and cleanup are outside the measured interval. Origin fetch and provider startup are excluded. Packet tracing is enabled identically for both implementations. Transport is local, so these fixtures measure no Internet latency or bandwidth savings.

For each destination, compare module initialization, HEAD, tracked-file lists, working-tree status, payload hash, and README presence against the baseline. Source status is checked after every call; the dirty-source scenario additionally checks the original edited file and excludes staged/untracked source files from the destination. Its selected commit is missing from the source reference, and the remote advances beyond that selected commit. Every initialized destination is checked for absence of persistent alternates. Cached single/sibling/nested/tiny cases make no object `want` requests with the optimization. The existing source-deletion/fsck tests remain separate coverage.

Active selection, update=none, and fetchJobs scenarios retain their selected/skipped modules. In the unreachable-module case, healthy siblings are cloned as with native Git, but Git may abort before checking out their files; successful worktree creation is not a claim of complete submodule population after a failure. All 120 calls matched the corresponding baseline state in these checks.

Times below are milliseconds, listed in execution order within each implementation. Means and percentages above use unrounded samples. Five observations describe these local runs; they do not establish a universal speedup or statistical significance for small differences.

| Scenario | Before: all five samples (ms) | After: all five samples (ms) |
|---|---|---|
| 1 cached submodule | 144.7, 117.7, 118.3, 116.3, 115.4 | 111.5, 104.6, 96.4, 94.5, 95.5 |
| 4 cached sibling submodules | 374.3, 362.3, 377.4, 354.0, 584.2 | 330.1, 328.8, 333.2, 320.8, 314.8 |
| 2 parents + 2 nested submodules | 348.9, 346.6, 348.3, 343.2, 349.3 | 302.4, 305.0, 305.8, 315.4, 303.2 |
| 12 tiny cached submodules | 235.2, 248.4, 239.3, 244.7, 234.7 | 396.9, 407.0, 398.6, 420.0, 400.2 |
| Active pathspec: 2 of 4 selected | 185.0, 200.2, 255.2, 196.5, 189.5 | 194.7, 191.5, 193.6, 198.0, 187.7 |
| Literal active=true pathspec | 112.2, 108.1, 110.0, 104.4, 111.2 | 108.8, 116.5, 117.0, 121.9, 119.5 |
| update=none: 1 of 4 skipped | 258.3, 262.8, 271.5, 275.4, 269.6 | 271.6, 271.0, 272.5, 274.3, 335.1 |
| Custom fetchJobs=4 | 153.8, 157.9, 161.3, 163.4, 499.7 | 167.2, 417.0, 168.9, 166.4, 219.8 |
| Stale + dirty source, 4 modules | 333.7, 318.4, 325.1, 310.2, 308.2 | 272.8, 280.9, 268.5, 261.1, 256.6 |
| Shallow sources, 4 modules | 306.6, 306.1, 311.1, 309.6, 308.1 | 411.2, 409.7, 419.5, 512.1, 413.7 |
| Uninitialized sources, 4 modules | 309.0, 308.5, 300.8, 361.5, 355.7 | 354.8, 345.2, 380.0, 390.5, 387.0 |
| Unreachable first module + 4 healthy siblings | 332.9, 326.2, 321.9, 326.7, 325.0 | 351.0, 350.3, 345.7, 345.3, 346.0 |

</details>

### Regression coverage

- **65 tests pass** in `GitVcsDriverCore.test.ts` on Node 24.20.0. The cached fixture uses two siblings and a nested submodule over `file://` transport; packet tracing asserts no object `want` requests when the matching source repositories already contain all objects. This assertion fails on the original PR implementation (`462f597`) and passes with the correction.
- Fixtures cover stale/populated, shallow and uninitialized sources; spaces and differing submodule names; a caller in a project subdirectory; a selected commit missing from the source; and a remote newer than the selected commit.
- Staged, unstaged and untracked source changes remain untouched and are not copied into the destination. Destination submodules retain the selected gitlinks and pass `fsck` and `reset --hard` after source object stores are deleted.
- Active-submodule selection (including the literal `submodule.active=true` pathspec) and `update=none` keep their ordinary Git behavior. An unreachable first submodule still allows Git to clone healthy siblings. Both added edge-case tests fail on `57a6e533d`, pass against the original baseline, and pass with the correction. The existing best-effort checkout-failure test remains passing.
- Targeted lint, formatting, server-only typecheck and `git diff --check` pass. Typecheck reports existing unrelated Effect suggestions, with no errors. No correctness regression was observed in this coverage. The edge-case measurements above record both speedups and timing regressions.

## Checklist

- [x] This PR is small and focused
- [x] I explained what changed and why
- [x] Focused behavioral tests and before/after measurements are included

Model and harness: GPT-6 / Codex.

---

> [!NOTE]
> **Medium Risk**
> Changes Git submodule initialization during worktree creation with new branching and fallbacks; incorrect reference handling could yield wrong commits or slow paths, though failures still degrade to the prior recursive update behavior.
> 
> **Overview**
> **Worktree creation** no longer relies on a plain recursive `git submodule update` when the caller’s checkout already has submodule repos locally. After `git worktree add`, the driver resolves the source repo top-level from `input.cwd` and runs new **`prepareSubmodules`**, which walks each submodule path, clones with **`--reference`** on the matching source checkout when a `.git` exists there, **`--dissociate`** so the new worktree owns its objects, then recurses for nested submodules.
> 
> Repos with non-default **`submodule.active` / `update` / parallel fetch** settings, missing references, or failed reference clones still use the original **`submodule update --init --recursive`** path (with the same best-effort warning if checkout fails).
> 
> **Tests** replace one happy-path case with a parametrized matrix (cached, populated, shallow, uninitialized, inactive, unreachable, `update=none`, etc.) covering nested submodules, pinned gitlinks, independence after deleting source object stores, untouched source dirty state, and packet tracing for the fully cached clone case.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 1001b4cc06ac0679c857f00bc1ab7f4bded51d2f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Reuse local source submodule objects when creating worktrees in `GitVcsDriver`
> Adds a recursive `prepareSubmodules` helper that clones each configured direct child submodule from the matching source submodule repository with dissociated objects, then recurses into nested children the same way. `createWorktree` now resolves the top-level source repository from the caller's working directory and invokes this helper before falling back to the ordinary recursive `git submodule update`.
> - When custom active/fetch-jobs/update policies are set, or when repository/gitmodules inspection fails, the code falls back to ordinary recursive `git submodule update` for the whole destination.
> - Individual per-submodule reference-clone failures fall back to a normal update for that child only; the overall worktree creation still completes with a warning if both population paths fail.
> - Expands the integration test suite in [GitVcsDriverCore.test.ts](https://github.com/pingdotgg/t3code/pull/9854/files#diff-ac30488287dac9aa8d4cb819c508d77d2e47267d4aae35819e6fa24b86312c26) to eight source-state scenarios (cached, populated, shallow, uninitialized, inactive, explicitly active, unreachable, update-disabled) with direct and nested submodules.
> - Risk: `prepareSubmodules` inspects `.gitmodules` and repo config to decide whether reference-based cloning is safe; if that inspection is wrong it could skip the optimization or produce a partial clone — verify behavior in [GitVcsDriverCore.ts](https://github.com/pingdotgg/t3code/pull/9854/files#diff-1c3e2b9763c6526006485633fb677c83dc54cc1ea5563a1e51aab707c35854d8) when custom `submodule.active` or `submodule.<name>.update` policies are present.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 1001b4c.</sup>
> 
>

### CODE REVIEW

File: apps/server/src/vcs/GitVcsDriverCore.ts [hunk 1]
```diff
@@ -2849,19 +2849,46 @@ export const makeGitVcsDriverCore = Effect.fn("makeGitVcsDriverCore")(function*
...
+      ).pipe(
+        Effect.map(splitNullSeparatedGitStdoutPaths),
+        Effect.orElseSucceed(() => []),
+      );
+      // Borrow only during cloning: removing or pruning the source must not
+      // break the new worktree. Git still fetches objects absent from references.
+      const referenceArgs = references.length
+        ? [
+            "--dissociate",
+            ...references.flatMap((ref) => ["--reference", path.resolve(input.cwd, ref)]),
```
[the remark this answers is not in the document]
Comment (Clowdyffs) [reply apps/server/src/vcs/GitVcsDriverCore.ts]: Addressed in 57a6e533d. Each direct child now receives only its matching source reference, then nested children are initialized with their corresponding source paths. Custom active/update and clone concurrency settings retain the ordinary recursive Git command.

Added a cached two-sibling plus nested-submodule fixture over file:// transport that asserts no object want requests. That assertion fails against 462f597 and passes with this correction; all 63 focused driver tests pass. The PR body now includes fresh measurements of the revised implementation and explicitly separates single-submodule timing from multi-submodule reuse coverage.

**Comment by Clowdyffs** [comment]
@codex review

The repeated-reference finding is addressed in 57a6e533d, including a transport-level regression test that fails on the previous implementation. This addresses the same object-reuse coverage gap raised in Macroscope’s approvability comment. The PR description has updated measurements and regression coverage for the revised implementation.

File: apps/server/src/vcs/GitVcsDriverCore.ts [hunk 2]
```diff
@@ -2831,6 +2831,79 @@ export const makeGitVcsDriverCore = Effect.fn("makeGitVcsDriverCore")(function*
...
+      const configs = [config, gitmodules];
+      const entries = configs.flatMap(splitNullSeparatedGitStdoutPaths);
+      // Explicit paths override Git's active/update policies and parallel cloning.
+      // Leave custom policies to the ordinary command rather than reimplement them.
+      if (
+        configs.some((result) => result.stdoutTruncated || result.exitCode > 1) ||
+        entries.some(
+          (entry) =>
+            /^submodule\.(active|fetchjobs|.*\.(active|update))\n/.test(entry) &&
+            !entry.endsWith(".active\ntrue"),
```
[the remark this answers is not in the document]
Comment (Clowdyffs) [reply apps/server/src/vcs/GitVcsDriverCore.ts]: Addressed in 1001b4cc0. The exemption now matches only submodule.<name>.active=true, so the global submodule.active=true pathspec uses ordinary Git selection. The new inactive-true fixture fails on 57a6e533d and passes on both the original baseline and the correction.

File: apps/server/src/vcs/GitVcsDriverCore.ts [hunk 3]
```diff
@@ -2831,6 +2831,79 @@ export const makeGitVcsDriverCore = Effect.fn("makeGitVcsDriverCore")(function*
...
+          reference,
+          "--dissociate",
+          "--",
+          submodule,
+        ]).pipe(Effect.catch(() => runGit("GitVcsDriver.cloneSubmodule", destination, args)));
+      } else {
+        yield* runGit("GitVcsDriver.cloneSubmodule", destination, args);
+      }
+      // Dissociation keeps the destination independent of source pruning/deletion.
+      yield* prepareSubmodules(reference, path.join(destination, submodule));
```
[the remark this answers is not in the document]
Comment (Clowdyffs) [reply apps/server/src/vcs/GitVcsDriverCore.ts:2903]: Addressed in 1001b4cc0 by retrying the ordinary recursive update if preparation fails. The new unreachable-first-submodule fixture verifies Git still clones healthy siblings; Git may abort before checking out their files after retrying the failed clone. This test fails on 57a6e533d and passes on both the original baseline and the correction. All 65 focused driver tests pass.

**Comment by Clowdyffs** [comment]
@codex review

1001b4cc0 addresses both new findings: the global active pathspec and continuing ordinary Git initialization after a failure. Both regression cases were checked against the original baseline and the reviewed commit; all 65 focused driver tests, targeted lint/formatting, and the server typecheck pass. The full production diff remains 98 changed lines.

**Comment by Clowdyffs** [comment]
Updated the [edge-case measurement matrix](https://github.com/pingdotgg/t3code/pull/9854#user-content-edge-case-measurements-on-the-final-commit) with explanations of the slower cases. Recalculated the published numbers from the raw samples; the existing means, percentages, and absolute differences were correct. No code changed (`1001b4cc0`).

- Twelve tiny modules add **164.1 ms** despite zero object requests: per-module Git invocations, reference setup, and dissociation outweigh transferring their tiny histories.
- Four shallow sources add **124.9 ms**: unsuitable reference attempts fall back to ordinary cloning, with no reduction in object requests.
- Four uninitialized sources add **44.4 ms**: reference checks and separate updates run despite having no initialized source checkouts to reuse.
- The immediately failing local-URL case adds **21.1 ms** before the ordinary recursive fallback. Slow remote failures can cost more.

These explanations follow the command paths and packet traces; timings do not isolate each subprocess's contribution. The 44–164 ms overhead is likely minor during first-message setup because it occurs once per newly created worktree, not on each subsequent message. User perception was not measured, and larger module counts or slower failures can increase the cost. The separate large-repository sample saved 15.82 seconds; the PR retains both the benefits and the negative rows so reviewers can assess the tradeoff.

All 120 measured calls across 12 scenarios matched the baseline destination state in the documented checks. The PR includes all five samples per implementation and identifies the synthetic local-transport measurements as Linux/btrfs results.
